### PR TITLE
RHIDP-2023: Added a procedure for running RHDH behing a corporate proxy

### PR DESCRIPTION
**Pull request by shipsing**
Purpose: To resolve this issue: https://issues.redhat.com/browse/RHIDP-2023 

Preview link: https://redhat-developer.github.io/red-hat-developers-documentation-rhdh/pr-225/admin-rhdh/#assembly-running-rhdh-behind-a-proxy

**Cherrypick version:** 1.2.x

SME: @rm3l 
QE: @albarbaro 
Peer: @linfraze

### CODE REVIEW

File: titles/admin-rhdh/title-admin.adoc
Comment (rm3l) [inline titles/admin-rhdh/title-admin.adoc:22]:
```suggestion
include::admin/proc-running-rhdh-behind-a-proxy.adoc[leveloffset=+1]
```

Is this the right path? I couldn't see this new section from the preview website:

image

**Review by rm3l — CHANGES_REQUESTED** [review]

File: titles/admin-rhdh/title-admin.adoc
Comment (shipsing) [reply titles/admin-rhdh/title-admin.adoc:22]: Updated, Thanks

File: admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (shipsing) [inline admin/proc-running-rhdh-behind-a-proxy.adoc:66]: @rm3l, I think we should not put this piece of information in a note because this is also a key step for configuring proxy information. Here, the user can make a decision whether they want to go with proxy configuration in a ConfigMap file or an individual CR file based on their requirements.

So, I have made this step part of the procedure and notified users to perform one of these two steps.

File: admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (shipsing) [inline admin/proc-running-rhdh-behind-a-proxy.adoc:73]: @rm3l, Could you please share a yaml example for setting proxy variables in a configmap file?

File: admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [reply admin/proc-running-rhdh-behind-a-proxy.adoc]: That makes sense. But maybe it is worth calling out that editing the operator default config is the responsibility of an administrator (with access to the operator namespace). So it can be a way for a cluster admin to apply the proxy settings to all the operator users.
The other alternative, at the individual CR level, can be done by regular users.

Comment (shipsing) [reply admin/proc-running-rhdh-behind-a-proxy.adoc]: Which role would configure Helm configuration?? adminstrators, developers, or both of them.

Comment (rm3l) [reply admin/proc-running-rhdh-behind-a-proxy.adoc]: For Helm, it is quite different. Each Helm-based instance is deployed from a user-provided `values` file. So a regular user would need to add the proxy settings to their `values` file(s).

File: admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [reply admin/proc-running-rhdh-behind-a-proxy.adoc]: Well, the default config of the operator is quite a large ConfigMap, which I'm attaching [here](https://github.com/redhat-developer/red-hat-developers-documentation-rhdh/files/15333866/rhdh-operator--backstage-default-config-1.2.yaml.txt) if you want, but it can be fetched by running the following command:
```shell
$ oc get configmap backstage-default-config -n rhdh-operator -o yaml > rhdh-operator-default-config.yaml
```

For your convenience, here is the diff with the proxy settings added to the init and main container of the RHDH application pod listed in the operator default config ConfigMap:
```diff
diff --git a/rhdh-operator--backstage-default-config-1.2.yaml b/rhdh-operator--backstage-default-config-1.2.yaml
index 99eea5d..30a7367 100644
--- a/rhdh-operator--backstage-default-config-1.2.yaml
+++ b/rhdh-operator--backstage-default-config-1.2.yaml
@@ -211,6 +211,12 @@ data:
               env:
                 - name: NPM_CONFIG_USERCONFIG
                   value: /opt/app-root/src/.npmrc.dynamic-plugins
+                - name: HTTP_PROXY
+                  value: '<http_proxy_url>'
+                - name: HTTPS_PROXY
+                  value: '<https_proxy_url>'
+                - name: NO_PROXY
+                  value: '<no_proxy_settings>'
               volumeMounts:
                 - mountPath: /dynamic-plugins-root
                   name: dynamic-plugins-root
@@ -264,6 +270,12 @@ data:
               env:
                 - name: APP_CONFIG_backend_listen_port
                   value: "7007"
+                - name: HTTP_PROXY
+                  value: '<http_proxy_url>'
+                - name: HTTPS_PROXY
+                  value: '<https_proxy_url>'
+                - name: NO_PROXY
+                  value: '<no_proxy_settings>'
               volumeMounts:
                 - mountPath: /opt/app-root/src/dynamic-plugins-root
                   name: dynamic-plugins-root
```

File: titles/admin-rhdh/title-admin.adoc
Comment (rm3l) [inline titles/admin-rhdh/title-admin.adoc:22]: Is the path fixed? I can still see the error on the preview website: `Unresolved directive in title-admin.adoc - include::admin/proc-running-rhdh-behind-a-proxy.adoc[leveloffset=+1]`
https://redhat-developer.github.io/red-hat-developers-documentation-rhdh/pr-225/admin-rhdh/

Comment (shipsing) [reply titles/admin-rhdh/title-admin.adoc]: The path is now fixed and you can view the procedure. Sorry, some content restructuring happened and i forgot to include the correct path.

File: admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (shipsing) [reply admin/proc-running-rhdh-behind-a-proxy.adoc]: I included all these details in the into para because providing these deatils in the middle of the procedure distracts the user from the procedure.

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:15]: > This configuration applies the proxy settings to all the users within the namespace where the Operator is deployed.

Just a small clarification on this as I think the "within the namespace ..." part can be misleading: I would say something like this: "This configuration applies the proxy settings to all the users of the operator.".

As the operator watches custom resources cluster-wide, the proxy settings configuration (once added by an admin to the operator default config) will be applied to all the users, regardless of the namespace in which they deploy their operator-based RHDH application (not necessarily within the namespace where the operator is deployed).

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:16]:
```suggestion
* As a developer, you can configure the proxy variables in a custom resource (CR) file. This configuration applies the proxy settings to the {product-very-short} application created from that CR.
```

In the end, the proxy settings configured by the developer are applied to the RHDH application.

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:72]:
```suggestion
.. Search for a ConfigMap file named `backstage-default-config` in the namespace where the Operator is deployed (`rhdh-operator` by default) and open it. 
```

The namespace where the RHDH operator is `rhdh-operator` by default. Should we add that here?

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:93]: There are 2 containers where the proxy settings need to be added. The one above is for the init container, but we also need to add them to the main application container.

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:83]: Even if the other fields are omitted, do you want to add the top-level fields as well, to make it clear what the path is? Something like this?
```yaml
# Other fields omitted
  deployment.yaml: |-
    apiVersion: apps/v1
    kind: Deployment
    spec:
      template:
        spec:
          initContainers:
            - name: install-dynamic-plugins
              env:
                - name: NPM_CONFIG_USERCONFIG
                  value: /opt/app-root/src/.npmrc.dynamic-plugins
                - name: HTTP_PROXY
                  value: 'http://10.10.10.105:3128'
                - name: HTTPS_PROXY
                  value: 'http://10.10.10.106:3128'
                - name: NO_PROXY
                  value: 'localhost,example.org'
          containers:
            - name: backstage-backend
              env:
                - name: APP_CONFIG_backend_listen_port
                  value: "7007"
                - name: HTTP_PROXY
                  value: 'http://10.10.10.105:3128'
                - name: HTTPS_PROXY
                  value: 'http://10.10.10.106:3128'
                - name: NO_PROXY
                  value: 'localhost,example.org'
```

File: modules/getting-started/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [inline modules/getting-started/proc-running-rhdh-behind-a-proxy.adoc:1]: Is this file used? If yes, maybe you can link to the Admin guide (which contains instructions for both developers and admins) instead, no?

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (shipsing) [reply modules/admin/proc-running-rhdh-behind-a-proxy.adoc:15]: Thanks, @rm3l, for the clarification and making me understand this part :)

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (shipsing) [reply modules/admin/proc-running-rhdh-behind-a-proxy.adoc:72]: Since we follow mod doc structure, we require to ensure that all the info we provide in a module is complete in itself. so, addint this info wud definitely be helpful.

File: modules/getting-started/proc-running-rhdh-behind-a-proxy.adoc
Comment (shipsing) [reply modules/getting-started/proc-running-rhdh-behind-a-proxy.adoc:1]: We do not need this file in GSG.

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:16]:
```suggestion
* As a developer, you can configure the proxy variables in a custom resource (CR) file. This configuration applies the proxy settings to the {product-very-short} application created from that CR.
```

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (rm3l) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:22]: Not sure what the second prerequisite means. We cannot create a project in Developer Hub, no?

**Review by rm3l — APPROVED** [review]
LGTM - thanks @shipsing!
But I guess, before merging your doc PR, we need to wait until https://github.com/janus-idp/backstage-showcase/pull/1225 is merged for the proxy env vars to be supported in RHDH.

**Review by albarbaro — APPROVED** [review]
All looks good to me! Thanks!
[feature pr](https://github.com/janus-idp/backstage-showcase/pull/1225) is not merged yet though, maybe we should wait for that first.

**Comment by rm3l** [comment]
@shipsing @albarbaro FYI, https://github.com/janus-idp/backstage-showcase/pull/1225 has been merged. You might need to wait until the downstream image is built for this feature to be available in RHDH.

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (linfraze) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:11]:
```suggestion
For Helm-based deployment, either a developer or a cluster administrator with permissions to create resources in the cluster can configure the proxy variables in a `values.yaml` Helm configuration file.
```

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (linfraze) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:21]:
```suggestion
* You have installed the {product} application. For more information, see xref:proc-install-rhdh_rhdh-getting-started[Administration guide for {product}].
```

Otherwise it shows up as _see [proc-install-rhdh_rhdh-getting-started]_ in the preview. 
Although I think you could also remove this xref entirely. It isn't really clear what the user is supposed to look at or gain from this guide.

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (linfraze) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:66]: I recommend splitting the Helm and Operator procedures into two separate proc modules at some point. Perhaps we create a separate ticket for this so we don't block this merge.

File: modules/admin/proc-running-rhdh-behind-a-proxy.adoc
Comment (linfraze) [inline modules/admin/proc-running-rhdh-behind-a-proxy.adoc:69]: I would consider creating separate proc modules for each of these role-based procedures also and specifying the role requirements in the Prerequisites section. However, that is an option we can address in the future.

File: titles/admin-rhdh/title-admin.adoc
Comment (linfraze) [inline titles/admin-rhdh/title-admin.adoc:22]: In the future, I would organize this into an assembly with multiple proc modules. For now, this is a great start to making this content available to users.

**Review by linfraze — APPROVED** [review]
lgtm - some suggestions and future considerations, but fine to merge

**Review by hmanwani-rh — APPROVED** [review]

**Pull request merged**
